src/plugins/intel_cpu/src/nodes/scaled_attn.cpp

```diff
@@ -806,6 +824,288 @@ bool ScaledDotProductAttention::isSupportedOperation(const std::shared_ptr<const
+            result[i] = cur[reverse_order[i]];
+        }
+        return result;
+    };
+    auto internal_mem_k = m_k_state->internal_state_mem();
+    auto internal_mem_v = m_v_state->internal_state_mem();
+
+    auto is_reset = m_k_state->is_reset_state();
+    if (is_reset) {
+        auto inputNumber = getOriginalInputsNumber();
+        auto&& init_graph_v_dims = getParentEdgeAt(inputNumber - 1)->getMemory().getStaticDims();
+        L0 = init_graph_v_dims.at(order[2]);
+        auto B_init = init_graph_v_dims.at(order[0]);
+        OPENVINO_ASSERT(B == B_init, "pastkv batch: ", B, " is not equal to batch of init subgraph: ", B_init);
+    } else if (internal_mem_k) {
+        auto block_desc = internal_mem_k->getDescWithType<BlockedMemoryDesc>();
+        auto&& dims = block_desc->getShape().getStaticDims();
+        auto B_state = dims[order[0]];
+        L0 = dims[order[2]];
+        OPENVINO_ASSERT(B == B_state, "pastkv batch: ", B, " is not equal to batch of state: ", B_state);
+    }
+    OPENVINO_ASSERT(B * (L0 + L1) > 0, "B or (L0+L1) is zero, B: ", B, ", L0: ", L0, ", L1: ", L1);
+    // resize buffer
+    if (B * H * (L0 + L1) * S > m_k_state->internal_state_max_size()) {
+        auto new_shape = {B, H, (L0 + L1) * 2, S};
+        auto mem_desc = std::make_shared<CpuBlockedMemoryDesc>(m_kvcache_precision,
+            Shape(reverse(new_shape)),
+            new_shape,
+            order);
+
+        auto new_internal_mem_k = std::make_shared<Memory>(getEngine(), mem_desc);
+        auto new_internal_mem_v = std::make_shared<Memory>(getEngine(), mem_desc);
+
+        PlainTensor new_pastk, new_pastv;
+        new_pastk.reset(new_internal_mem_k);
+        new_pastv.reset(new_internal_mem_v);
+        new_pastk = new_pastk.permute(order);
+        new_pastv = new_pastv.permute(order);
+        if (L0 > 0 && !is_reset) {
+            past_k.reset(internal_mem_k);
+            past_v.reset(internal_mem_v);
+            past_k = past_k.permute(order);
+            past_v = past_v.permute(order);
+            attn_memcpy(past_k, past_v, new_pastk, new_pastv);
+        }
+        internal_mem_k = new_internal_mem_k;
+        internal_mem_v = new_internal_mem_v;
+        past_k = new_pastk;
+        past_v = new_pastv;
+        m_k_state->assign_internal_state(new_internal_mem_k);
+        m_v_state->assign_internal_state(new_internal_mem_v);
+        m_k_state->assign_internal_state_max_size(B * H * (L0 + L1) * 2 * S);
+        m_v_state->assign_internal_state_max_size(B * H * (L0 + L1) * 2 * S);
+    }
+    auto new_shape = {B, H, (L0 + L1), S};
+    auto mem_desc = std::make_shared<CpuBlockedMemoryDesc>(m_kvcache_precision,
+        Shape(reverse(new_shape)),
+        new_shape,
+        order,
+        0,
+        VectorDims{},
+        internal_mem_k->getDescWithType<BlockedMemoryDesc>()->getStrides());
+    internal_mem_k->redefineDesc(mem_desc);
+    internal_mem_v->redefineDesc(mem_desc);
```

Comment:
we can see a comment above right `redefineDesc`'s declaration saying:
```
// Caution!!! This action invalidates the previous data layout. The old data may become unreachable.
```
But here we are relying on the assumption that `redefineDesc` to smaller size will not cause reallocation, this is dangerous, behavior may change in future.

Maybe we can use Memory with external pointer to interact with edge/state API, but shouldn't relying on `Memory` class to manage the memory.

Reply:
Currently depends on the behavior of `redefineDesc`, maybe will refactor in the future.